Use pre-compiled regular expressions

rtobar opened the pull request:
Some low-hanging fruit inspired by #76.

AlexWaygood reviewed (approved):
LGTM! This gets us most of the way to the speedup I reported in https://github.com/sphinx-contrib/sphinx-lint/issues/76#issuecomment-1755747859 :D

There's a few other things we can do to speed things up further, but I can tackle them in a followup PR tomorrow, since I've already got them around locally :)

AlexWaygood commented on sphinxlint/utils.py:
```diff
@@ -121,17 +121,21 @@ def looks_like_glued(match):
 
+_START_OF_COMMENT_BLOCK_PATTERN = re.compile(r"^\s*\.\.$")
+_PRODUCTION_LIST_DIRECTIVE_PATTERN = re.compile(r"^ *.. productionlist::")
+_COMMENT_PATTERN = re.compile(r"^ *\.\. ")
```
This one can also be used in the `hide_non_rst_blocks()` function immediately below `is_multline_non_rst_block()`

rtobar replied:
True! I hadn't realised because I was focusing on the output of the profiling, but you're absolutely right, I'll reuse it there too.

AlexWaygood commented:
A lot of precompiled regexes in this repo seem to be located in `sphinxlint.rst`; not sure if these should be as well. @hugovk, any thoughts on that?

hugovk commented:
> A lot of precompiled regexes in this repo seem to be located in `sphinxlint.rst`; not sure if these should be as well. @hugovk, any thoughts on that?

I think they're fine where they are, but no strong opinion either way :)

AlexWaygood commented:
> > A lot of precompiled regexes in this repo seem to be located in `sphinxlint.rst`; not sure if these should be as well. @hugovk, any thoughts on that?
> 
> I think they're fine where they are, but no strong opinion either way :)

Let's leave them where they are for now!

AlexWaygood reviewed (approved):
thanks!

rtobar commented:
Sorry, just added another one after another quick profiling, and renamed the `_PATTERN` objects to `_RE`, which is what the current codebase calls `re.Pattern` objects. I won't be adding anymore to this PR.

hugovk reviewed (approved):
Running with `main` on CPython docs:

```console
❯ time sphinx-lint Doc
No problems found.
sphinx-lint Doc  17.89s user 0.31s system 672% cpu 2.705 total
```
And this PR:
```console
❯ time sphinx-lint Doc
No problems found.
sphinx-lint Doc  13.98s user 0.30s system 641% cpu 2.225 total
```

A nice 18% improvement! Almost too easy!

hugovk commented:
With both this and https://github.com/sphinx-contrib/sphinx-lint/pull/79:

```console
❯ time sphinx-lint Doc
No problems found.
sphinx-lint Doc  11.93s user 0.31s system 626% cpu 1.954 total
```

27% improvement! :tada:
